doc/design/logging.md

```diff
@@ -0,0 +1,396 @@
+# Logging
+
+## Summary and problem description
+
+Logging is a crucial part of a framework since it informs the developer and
+user of the framework about anomalies and helps with debugging by providing
+contextual information.
+
+Additionally, when integrated into a separate framework, the log messages should
+be forwarded in order to have a single logging infrastructure.
+
+The logging should also be performant and not allocate memory. Ideally it should
+be possible to disable it at compile time and let the compiler optimize it away.
+
+It should also be possible to alter the behavior of the logging via environment
+variables, e.g. the log level.
+
+## Terminology
+
+### Log levels and how to use them
+
+| LogLevel | Usage |
+|----------|-------|
+| FATAL    | For fatal and non-recoverable errors which essentially lead to termination of the program. |
+| ERROR    | For severe but recoverable errors. |
+| WARN     | For cases when something unintentional happens which is not considered a severe error. |
+| INFO     | For important status information a user should be aware of. |
+| DEBUG    | Anything that is helpful for debugging. |
+| TRACE    | Any comment could be replaced by logging call with log level trace. |
+
+The log levels `FATAL`, `ERROR` and `WARN` should not be used independently but
```

Comment:
@elfenpiff @mossmaurice 

Please don't mix the logger with an error handler/reporter (I know an elf who heavily complained about this idea when it was proposed a few years ago ;) ). Since this is a library, it is totally fine to do this in my application

```cpp
int main() {
    IOX_LOG(ERROR) << "yadda yadda";
};
```

Please let the logger do what a logger is supposed to do, log messages.

I think we all agree that a call like `IOX_LOG(FATAL) << "yadda yadda";` or `IOX_LOG(ERROR) << "yadda yadda";` should terminate an application. At least I don't know of any logger which has such a behavior. This means we would need to do something like this

```cpp
IOX_LOG(FATAL, MY_ERROR_CODE) << "yadda yadda"; // does reports the error and returns
IOX_FOO(FATAL, MY_ERROR_CODE); // or at least IOX_FOO(FATAL); since in debug builds `ERROR` would also not return and e.g. lead to termination
```

or something like this

```cpp
IOX_LOG(WARN, MY_ERROR_CODE_1) << "yadda yadda";  // does report the error and returns
IOX_FOO(FATAL, MY_ERROR_CODE_2) << "yadda yadda";  // does report the error but does not return, e.g. terminate
```

and compare it so the proposal of @MatthiasKillat 

```cpp
IOX_RAISE(WARN, MY_ERROR_CODE_1) << "yadda yadda";  // does report the error and returns
IOX_RAISE(FATAL, MY_ERROR_CODE_2) << "yadda yadda";  // does report the error but does not return, e.g. terminate
```

It's unambiguous since it always reports the error and that is the important part. The user does not have to think which macro to use when he concludes that this is something which needs to be stored in the black box. This handler also ensures that the program does not continue with the execution if the error level is `FATAL`.

One could of course argue if `IOX_RAISE` should be split into two macros and I would also argue that this is a bad idea and a source for errors, e.g.

```
IOX_REPORT(MY_ERROR_CODE_1) << "yadda yadda";
IOX_FOO(FATAL); // or just nothing for WARN
```

- okay, it it is nothing for WARN, how do I know that the user just forgot to add the line?
- why is there a `IOX_FOO(FATAL)` but no `IOX_FOO(WARN)`?
- if there is a `IOX_FOO(WARN)`, `IOX_FOO(ERROR)` and `IOX_FOO(FATAL)` which must be used in combination with `IOX_REPORT(...)` why not combine it?
- can you list me all error codes with an error level of FATAL? With `IOX_RAISE(FATAL, ...)` it is just a `grep -irn "IOX_RAISE(FATAL"`.

The error handler should be the lowest building block and the error handler macro can use the logger to do something like
```
IOX_RAISE(FATAL, MY_ERROR) << "yadda yadda".
```

The reason for this is that the logger is more complex than the error handler. The logger must do something with the message (console, file) but the error handler/reporter must in the most naive case only terminate. If the error handler/reporter needs to send the error code via e.g. network it must still handle cases where an non-recoverable error occurs, and therefore it must be lower level than the logger.

@mossmaurice since I always assume that the user of my software is a psychopath who knows where I'm living, I would keep the error code. Without the error code one needs a mapping of file, line, fuction to a specific error and this for each release of iceoryx. It's also not fun to test for these errors.

Reply:
Assuming that I understood everything correctly.

1.) I agree that we should not split logging and reporting in fatal error cases. I.e. if there is a fatal error, only use `IOX_RAISE(FATAL)` or some variation which also logs, optionally executes some code and terminates.
2.) For non-fatal errors I would argue it is the same, i.e. no split of logging and error reporting. The difference is that now the program continues (likely/necessarily on a different path).
3.) `IOX_LOG(<NON-FATAL>)` would be used if there is no error but it is useful to log some information, like that some (non-error) event happened, how much time some call took etc.

 This means we only explicitly log good or at least neutral events, while the bad ones require reporting an error and logging happens automatically.